numpy/random/mtrand/randint_helpers.pxi.in

```diff
@@ -51,27 +57,61 @@ def _rand_{{npy_dt}}(low, high, size, rngstate):
     -------
     out : python integer or ndarray of np.{{np_dt}}
           `size`-shaped array of random integers from the appropriate
-          distribution, or a single such random int if `size` not provided.
+          distribution, or an array corresponding to the shapes of
+          `low` and `high`.
 
     """
-    cdef npy_{{npy_udt}} off, rng, buf
-    cdef npy_{{npy_udt}} *out
-    cdef ndarray array "arrayObject"
-    cdef npy_intp cnt
     cdef rk_state *state = <rk_state *>PyCapsule_GetPointer(rngstate, NULL)
+    cdef ndarray array "arrayObject"
+    cdef npy_{{npy_udt}} off, rng, buf
+    cdef npy_{{npy_udt}} *array_data
+
+    cdef ndarray lo, hi
+    cdef npy_intp i, cnt
+    cdef broadcast multi
+
+    lo = <ndarray>np.array(low)
+    hi = <ndarray>np.array(high)
 
-    rng = <npy_{{npy_udt}}>(high - low)
-    off = <npy_{{npy_udt}}>(<npy_{{npy_dt}}>low)
+    if lo.shape == hi.shape == ():
+        low = int(low)
+        high = int(high)
```

Comment:
I think the fix there is better, because it doesn't incur the cost of producing an arbitrary-precision integer.

Also, presumably having `int` there prevents that being folded into the nogil clause below